events/handlers.js

```diff
@@ -0,0 +1,184 @@
+          req.query.location == data.location
+        ) {
+          results.push(data);
+        }
+      })
+      .on("end", () => {
+        res.send(results);
+      });
+  });
+}
+
+function createEvent(req, res) {
+  lockEvents(req, res, (req, res) => {
+    // just add at the end of the file
+    fs.open(filePath, "a", 666, function (e, id) {
+      // convert JSON array to CSV string
+      converter.json2csv(
+        req.body,
+        (err, csv) => {
+          if (err) {
+            res.status(500).send();
+            return;
+          }
+          const row = "\n" + csv;
+          fs.write(id, row, null, "utf8", function () {
+            fs.close(id, function () {
+              res.status(200).send();
+            });
+          });
+        },
+        { prependHeader: false }
+      );
+    });
+  });
+}
+
+function updateEvent(req, res) {
+  lockEvents(req, res, (req, res) => {
+    let results = [];
+    let found = false;
+    fs.createReadStream(filePath)
+      .pipe(csv())
+      .on("data", (data) => {
+        if (data.id == req.params.eventId) {
+          found = true;
+          results.push(req.body);
+        } else {
+          results.push(data);
+        }
+      })
+      .on("end", () => {
+        if (found === false) {
+          res.status(404).send("event not found");
+          return;
+        }
+        const csvWriter = createCsvWriter({
+          path: filePath,
+          header: headerFile,
+        });
+        csvWriter.writeRecords(results).then(() => {
+          res.status(200).send();
+        });
+      });
+  });
+}
+
+function deleteEvent(req, res) {
+  lockEvents(req, res, (req, res) => {
+    let results = [];
+    let found = false;
+    fs.createReadStream(filePath)
+      .pipe(csv())
+      .on("data", (data) => {
+        if (data.id == req.params.eventId) {
+          found = true;
+          return;
+        }
+        results.push(data);
+      })
+      .on("end", () => {
+        if (found === false) {
+          res.status(404).send("event not found");
+          return;
+        }
+        const csvWriter = createCsvWriter({
+          path: filePath,
+          header: headerFile,
+        });
+        csvWriter.writeRecords(results).then(() => {
+          console.log("The CSV file was written successfully");
+          res.status(200).send();
+        });
+      });
+  });
+}
+
+function getEventsBatch(req, res) {
+  lockEvents(req, res, (req, res) => {
+    let results = [];
+    fs.createReadStream(filePath)
+      .pipe(csv())
+      .on("data", (data) => {
+        results.push(data);
+      })
+      .on("end", () => {
+        res.send(results);
+      });
+  });
+}
+
+function lockEvents(req, res, callback) {
+  lockFile.lock("some-file.lock", {}, function (er) {
+    if (er) {
+      console.error("error while locking file", er);
+      res.status(500).send();
+      return;
+    }
+    callback(req, res);
+    lockFile.unlock("some-file.lock", function (er) {
```

Comment:
We might run this `unlock` before we actually run all commands in our `callback` function. 
For example inside callback we can have `fs.write(id, row, null, "utf8", writeCallback)`. And  we are not going to run `writeCallback` right away because it is in async loop. But our `lockFile.unlock` is just next synchronous command which can be called while we are waiting for write to finish.
It might make sense to log/debug this flow for better understanding.

I'd imagine that it should be implemented like this:
```
callback(req, res, function() {
  lockFile.unlock("some-file.lock" ....
 });

lockEvents(req, res, (req, res, unlockCallback) => {
    fs.write(..., function () {
        ...
        unlockCallback();
    })
})
```

Or using `await` if we'd use promises
```
await lockFile.lock()
await callback()
await lockFile.unlock()
```